### PR TITLE
[GlobalIsel] combine extract vector element

### PR DESCRIPTION
**Pull request by tschuett**
scalarize compares

extelt (cmp X, Y), Index --> cmp (extelt X, Index),
                                 (extelt Y, Index)

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp [hunk 1]
```diff
@@ -453,3 +453,162 @@ bool CombinerHelper::matchInsertVectorElementOOB(MachineInstr &MI,
...
+                                        unsigned Depth) {
+  assert(MRI.getType(Src).isVector() && "expected a vector as input");
+  if (Depth == 2)
+    return false;
+
+  // We use the look through variant for higher hit rate and to increase the
+  // likelyhood of constant folding. The actual value is ignored. We only test
+  // *whether* there is a constant.
+
+  MachineInstr *SrcMI = getDefIgnoringCopies(Src, MRI);
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperVectorOps.cpp:467]: But this should be taken care of by copy->copy folding? I think we should mostly be removing all the getDefIgnoringCopies calls

**Review by arsenm — COMMENTED** [review]
SelectionDAG seems to not do this combine, and InstCombine does. Is this pattern coming from somewhere in legalization?

**Comment by tschuett** [comment]
It comes out of the middle-end. It is a combine to free the vector register by an instruction.

**Comment by tschuett** [comment]
https://github.com/llvm/llvm-project/blob/61d4ca872215d3dfff0b3c92151dcbdc546a0aab/llvm/test/CodeGen/AArch64/extract-vector-elt.ll#L483

The tricky case with an ext in between.

**Comment by tschuett** [comment]
We should add a Combiner.rst:
* Each combine is a precise pattern in Combine.td, for compile-time and targets can pick their choice of combines
* There are no visit functions
* Document the differences of the Gisel Combiner and InstCombine, i.e., the kind of combines and aggressiveness
* The Gisel combiner is a pure legalization artefact combiner?

The DAGCombiner seems to do more than combining legalization artefacts.

**Comment by tschuett** [comment]
From the Gisel documentation:
```
Combiner

Replaces patterns of instructions with a better alternative. Typically, this means improving run time performance by replacing instructions with faster alternatives but Combiners can also focus on code size or other metrics.
```

**Comment by tschuett** [comment]
That is fine by me. As long as we document the combiner and what kind of combines it has.

**Comment by aemerson** [comment]
Filed https://github.com/llvm/llvm-project/issues/92309 to remind us to do this

**Pull request closed without merging**